### PR TITLE
CC3 

### PR DESCRIPTION
**Pull request by zw0930**
## Description
CC3 ground state energy.

## Todos
  - [x] Add functions for CC3 t-equations.
  - [x] Test and check the results with Psi4.
  - [x] Merge the single-precision branch if that pull request will be merged first.

## Status
- [x] Ready to go

### CODE REVIEW

File: pycc/ccwfn.py [hunk 1]
```diff
@@ -305,10 +307,43 @@ def residuals(self, F, t1, t2):
 
         r1 = self.r_T1(o, v, F, ERI, L, t1, t2, Fae, Fme, Fmi)
         r2 = self.r_T2(o, v, F, ERI, L, t1, t2, Fae, Fme, Fmi, Wmnij, Wmbej, Wmbje, Zmbij)
-       
+             
         if isinstance(Fae, torch.Tensor):
-            del Fae, Fmi, Fme, Wmnij, Wmbej, Wmbje, Zmbij 
-
+            del Fae, Fmi, Wmnij, Wmbej, Wmbje, Zmbij
```
Comment (bgpeyton) [inline pycc/ccwfn.py:312]: what happens to `Fme` if `Fae` is a `torch.Tensor`, but the model is something other than `CC3`?

Comment (zw0930) [reply pycc/ccwfn.py:312]: Yes, that should be changed. Thanks for catching that, Ben.

**Review by bgpeyton — COMMENTED** [review]
could we add one more test to `test_031_cc3_ecc.py`? a slightly bigger basis set would be good. if it takes too much time, you could use H2 instead of H2O.

**Comment by zw0930** [comment]
> could we add one more test to `test_031_cc3_ecc.py`? a slightly bigger basis set would be good. if it takes too much time, you could use H2 instead of H2O.

Yes, how about H2 with aug-cc-pVDZ (or aug-cc-pVTZ). It will not take long (tested on my laptop). There's also H2O with cc-pVDZ commented out in the test that passed on my laptop. It might take a bit longer but still pretty quick.

**Review by bgpeyton — APPROVED** [review]
H2 with aDZ sounds fine. LGTM! thanks Zhe!

File: pycc/ccwfn.py [hunk 2]
```diff
@@ -555,6 +591,120 @@ def r_T2(self, o, v, F, ERI, L, t1, t2, Fae, Fme, Fmi, Wmnij, Wmbej, Wmbje, Zmbi
...
+        tmp = 2 * t3 - t3.swapaxes(1,2) - t3.swapaxes(0,2)
+        S = contract('abc,dbc->ad', tmp, Wamef.swapaxes(0,1)[k])
+        return S
+
+    # t_ijkabc -> - Silab
+    def sigma_T2_C(self, j, k, l, Wmnie, t3):
+        contract = self.contract
+        tmp = 2 * t3 - t3.swapaxes(1,2) - t3.swapaxes(0,2)
+        S = contract('abc,c->ab', tmp, Wmnie[j,k,l])
+        return (-1) * S
```
[the remark this answers is not in the document]
Comment (zw0930) [reply pycc/ccwfn.py:707]: Oh I see. The loop over l should be evitable. I'll make the modification. Thank you for the comment!

File: pycc/ccwfn.py [hunk 3]
```diff
@@ -333,10 +336,42 @@ def residuals(self, F, t1, t2):
...
+                        r2[i,j] += tmp
+                        r2[j,i] += tmp.swapaxes(0,1)                      
+                        # t_ijkabc -> Sijad
+                        tmp = self.sigma_T2_B(k, Wamef_cc3, t3)
+                        r2[i,j] += tmp
+                        r2[j,i] += tmp.swapaxes(0,1)
+                        # t_ijkabc -> Silab
+                        tmp = self.sigma_T2_C(j, k, Wmnie_cc3, t3)
+                        r2[i] += tmp
+                        r2[:,i] += tmp.swapaxes(1,2)
```
Comment (lothian) [inline pycc/ccwfn.py:370]: My next question is whether it's necessary to handle these terms as separate function calls.  Again, I refer to the comparable code in `t_vikings`, where the contractions appear simpler and thus a bit more readable:

`X1[i] += contract('abc,bc->a',(t3 - t3.swapaxes(0,2)), L[j,k,v,v])`
`X2[i,j] += contract('abc,c->ab',(t3 - t3.swapaxes(0,2)), F[k,v])`
`X2[i,j] += contract('abc,dbc->ad', (2.0*t3 - t3.swapaxes(1,2) - t3.swapaxes(0,2)),ERI[v,k,v,v])`
`X2[i] -= contract('abc,lc->lab', (2.0*t3 - t3.swapaxes(1,2) - t3.swapaxes(0,2)),ERI[j,k,o,v])`

File: pycc/ccwfn.py [hunk 4]
```diff
@@ -555,7 +590,91 @@ def r_T2(self, o, v, F, ERI, L, t1, t2, Fae, Fme, Fmi, Wmnij, Wmbej, Wmbje, Zmbi
...
+    def build_cc3_Wamef(self, o, v, ERI, t1):
+        contract = self.contract
+        if isinstance(t1, torch.Tensor):
+            W = ERI[v,o,v,v].clone()
+        else:
+            W = ERI[v,o,v,v].copy()
+        W = W - contract('na,nmef->amef', t1, ERI[o,o,v,v])
+        return W   
+
+    def build_cc3_Wabei(self, o, v, ERI, t1):
```
Comment (lothian) [inline pycc/ccwfn.py:636]: I'm not going to request changes to this function, but I do think we should re-visit it at some point.  I think the structure you've adopted is intended to follow that from the corresponding functions in Psi4.  However, many of the algorithmic issues that are important in that C++ code don't really apply to a Python variant, and so we might be able to simply this section and make it more readable.

**Review by lothian — APPROVED** [review]

File: pycc/ccwfn.py (hunk 4 above)
Comment (zw0930) [reply pycc/ccwfn.py:636]: That makes sense. I will keep it in mind and think about it.

**Pull request merged**